### PR TITLE
WIP Add support for snaps

### PR DESCRIPTION
**Pull request by timClicks**

### CODE REVIEW

File: snap.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+	err := snap.readProxySettingsFromSnapSystemConfig(&res)
+	if err != nil {
+		return proxy.Settings{}, err
+	}
+	if res != (proxy.Settings{}) {
+		return res, nil
+	}
+
+	err = snap.readEtcEnvironmentForProxySettings(&res)
```
Comment (Veebers) [inline snap.go]: As above, we don't care about the system proxy settings, just those set for snapd.

File: snap.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	return res, nil
+}
+
+func (snap *snap) readProxySettingsFromSnapSystemConfig(settings *proxy.Settings) error {
+	configKeys := []string{"http", "https", "ftp"}
+
+	for _, proxyType := range configKeys {
+		// for snap <2.3.6, this should be "core"
+		cmd := fmt.Sprintf("snap get system proxy.%s", proxyType)
+		out, _, err := RunCommandWithRetry(cmd, nil)
```
Comment (Veebers) [inline snap.go]: I'm pretty sure if there is no proxy set this will error (```snap get``` return non-zero)

File: snap.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+// AddRepository is defined on the PackageManager interface.
+func (snap *snap) AddRepository(repo string) error {
+	return errors.New("not implemented")
+}
+
+// RemoveRepository is defined on the PackageManager interface.
+func (snap *snap) RemoveRepository(repo string) error {
+	return errors.New("not implemented")
```
Comment (Veebers) [inline snap.go]: We shouldn't be returning an error, perhaps just a log message and return nil.

File: commands/snap.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// Copyright 2019 Canonical Ltd.
+// Copyright 2015 Cloudbase Solutions SRL
```
Comment (Veebers) [inline commands/snap.go]: Is this license needed?

File: commands/snap.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+var unsetProxy = `
+if grep -qs ^http_proxy= /etc/environment; then
+	sed -i 's/^http_proxy=.*//' /etc/environment
+fi
+
+if grep -qs ^https_proxy= /etc/environment; then
+	sed -i 's/^https_proxy=.*//' /etc/environment
+fi
+`[1:]
```
[the remark this answers is not in the document]
Comment (timClicks) [reply commands/snap.go]: Good point. Will make sure that I keep setting to internal snap stuff..

File: commands/snap.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+// snapCmder provides commands that are relevant for snap-based systems.
+var snapCmder = packageCommander{
+	update:        "snap refresh",
+	upgrade:       `snap refresh %s`,
+	listInstalled: "snap list | cut -d ' ' -f 1 | tail -n+2",
+	install:       "snap install",
+	cleanup:       "snap refresh",
+	listAvailable: "snap list",
```
[the remark this answers is not in the document]
Comment (timClicks) [reply commands/snap.go]: Good point

File: commands/snap.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+if grep -qs ^https_proxy= /etc/environment; then
+	sed -i 's/^https_proxy=.*//' /etc/environment
+fi
+`[1:]
+
+// snapCmder provides commands that are relevant for snap-based systems.
+var snapCmder = packageCommander{
+	update:        "snap refresh",
+	upgrade:       `snap refresh %s`,
+	listInstalled: "snap list | cut -d ' ' -f 1 | tail -n+2",
```
[the remark this answers is not in the document]
Comment (timClicks) [reply commands/snap.go]: Turns out that this isn't used by the manager later on.. will remove

File: commands/snap.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	install:       "snap install",
+	cleanup:       "snap refresh",
```
[the remark this answers is not in the document]
Comment (timClicks) [reply commands/snap.go]: Sounds good

File: commands/snap.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	sed -i 's/^http_proxy=.*//' /etc/environment
+fi
+
+if grep -qs ^https_proxy= /etc/environment; then
+	sed -i 's/^https_proxy=.*//' /etc/environment
+fi
+`[1:]
+
+// snapCmder provides commands that are relevant for snap-based systems.
+var snapCmder = packageCommander{
```
[the remark this answers is not in the document]
Comment (timClicks) [reply commands/snap.go]: In my defence, was going with the project's conventions:

```go
// NewAptPackageCommander returns a PackageCommander for apt-based systems.
func NewAptPackageCommander() PackageCommander {
	return &aptCmder
}

// NewYumPackageCommander returns a PackageCommander for yum-based systems.
func NewYumPackageCommander() PackageCommander {
	return &yumCmder
}

// NewZypperPackageCommander returns a PackageCommander for zypper-based systems.
func NewZypperPackageCommander() PackageCommander {
	return &zypperCmder
}

// NewSnapPackageCommander returns a PackageCommander for snap-supported systems.
func NewSnapPackageCommander() PackageCommander {
	return &snapCmder
}

```

File: commands/snap.go (hunk 4 above)
Comment (timClicks) [reply commands/snap.go]: I'm not sure. It's included in the apt file and I've copied it straight from there. So I've assumed that there are still elements of the original copyright that have been preserved.

File: commands/snap.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (timClicks) [reply commands/snap.go]: Setting the proxy is wholly within `manager/snap.go`, so I've removed this bash code.

File: commands/snap.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (timClicks) [reply commands/snap.go]: Removed

**Comment by jameinel** [comment]
I assume this landed in a different PR, if not, we can review this again.

**Pull request closed without merging**